### PR TITLE
Adjust MetadataListing error code based on cause

### PR DESCRIPTION
**Pull request by pajaks**
## Description

Find exception with error code in stack trace and use it in new TrinoException.

For example. In case stack trace looks like:
```
java.lang.RuntimeException
   io.trino.spi.TrinoException (USER_ERROR)

```
MetadataListening was throwing `GENERIC_INTERNAL_ERROR` instead `USER_ERROR`. After this PR error code will be searching stack trace for error code if present.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by pajaks** [comment]
Ci hit: https://github.com/trinodb/trino/issues/18697

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/ExceptionUtil.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    private ExceptionUtil() {}
+
+    public static ErrorCodeSupplier findErrorCode(Throwable throwable)
+    {
+        return getCausalChain(throwable).stream()
+                .filter(TrinoException.class::isInstance)
+                .map(TrinoException.class::cast)
+                .findFirst()
+                .map(t -> (ErrorCodeSupplier) t::getErrorCode)
+                .orElse(GENERIC_INTERNAL_ERROR);
```
Comment (findinpath) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/ExceptionUtil.java:33]: I'd argue that returning  `Optional<ErrorCodeSupplier>` is more appropriate.

This is something that the user of this method should rather decide.
```
.orElse(GENERIC_INTERNAL_ERROR);
```

File: core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java [hunk 2]
```diff
@@ -349,12 +348,8 @@ private static Map<SchemaTableName, List<ColumnMetadata>> doListTableColumns(Ses
 
     private static TrinoException handleListingException(RuntimeException exception, String type, String catalogName)
     {
-        ErrorCodeSupplier result = GENERIC_INTERNAL_ERROR;
-        if (exception instanceof TrinoException trinoException) {
-            result = trinoException::getErrorCode;
-        }
         return new TrinoException(
-                result,
+                findErrorCode(exception),
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java:352]: Add in the description of this PR the stacktrace which motivated you to make this change.

**Review by findinpath — APPROVED** [review]

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/ExceptionUtil.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+package io.trino.plugin.base.util;
+
+import io.trino.spi.ErrorCodeSupplier;
+import io.trino.spi.TrinoException;
+
+import java.util.Optional;
+
+import static com.google.common.base.Throwables.getCausalChain;
+
+public final class ExceptionUtil
```
Comment (ebyhr) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/ExceptionUtil.java:23]: https://github.com/trinodb/trino/pull/21210 is going to introduce `Exceptions` class. Let's use the same name.

File: lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/ExceptionsTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.Optional;
+
+import static io.trino.plugin.base.util.Exceptions.findErrorCode;
+import static io.trino.spi.StandardErrorCode.TOO_MANY_REQUESTS_FAILED;
+import static io.trino.spi.StandardErrorCode.USER_CANCELED;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ExceptionsTest
```
Comment (ebyhr) [inline lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/ExceptionsTest.java:27]: Please add `final` and rename to `TestExceptions`. 

https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/develop/tests.md#conventions-and-recommendations
> * Test classes should be defined as package-private and final.
> * Test class names must start with Test, for example TestExample

File: lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/ExceptionsTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    void testFindErrorCodeWithExceptionStack()
+    {
+        Throwable exception1 = new TrinoException(USER_CANCELED, "fake exception");
+        Throwable exception2 = new RuntimeException(exception1);
+        Optional<ErrorCodeSupplier> errorCodeSupplier = findErrorCode(exception2);
+        assertThat(errorCodeSupplier.orElseThrow().toErrorCode()).isEqualTo(USER_CANCELED.toErrorCode());
+
+        Throwable exception3 = new RuntimeException(exception2);
+        errorCodeSupplier = findErrorCode(exception3);
```
Comment (ebyhr) [inline lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/ExceptionsTest.java:54]: I would inline instead of reusing the variable.

**Comment by pajaks** [comment]
Closing as generic resolution seems to be excessive.

**Pull request closed without merging**